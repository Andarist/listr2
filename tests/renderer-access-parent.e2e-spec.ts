import delay from 'delay'
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root'

describe('show subtasks', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // nvFwliir81dQSzoa7xPXaq0E1lauH674
  it('should be able to change the title of parent from subtask', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): Listr =>
            task.newListr((parent) => [
              {
                title: 'This is a subtask.',
                task: async (): Promise<void> => {
                  await delay(20)
                  parent.title = 'Title changed from subtask.'
                }
              },

              {
                title: 'This is a subtask.',
                task: async (): Promise<void> => {
                  await delay(30)
                }
              }
            ])
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('nvFwliir81dQSzoa7xPXaq0E1lauH674-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('nvFwliir81dQSzoa7xPXaq0E1lauH674-err')
    expect(mockExit.mock.calls).toMatchSnapshot('nvFwliir81dQSzoa7xPXaq0E1lauH674-exit')
  })
})
