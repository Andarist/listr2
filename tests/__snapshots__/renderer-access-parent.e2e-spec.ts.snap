// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show subtasks should be able to change the title of parent from subtask: nvFwliir81dQSzoa7xPXaq0E1lauH674-err 1`] = `[]`;

exports[`show subtasks should be able to change the title of parent from subtask: nvFwliir81dQSzoa7xPXaq0E1lauH674-exit 1`] = `[]`;

exports[`show subtasks should be able to change the title of parent from subtask: nvFwliir81dQSzoa7xPXaq0E1lauH674-out 1`] = `
[
  [
    "◼ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G❯ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G❯ This task will execute.
  ◼ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G❯ This task will execute.
  ◼ This is a subtask.
  ◼ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G❯ This task will execute.
  ❯ This is a subtask.
  ◼ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G❯ Title changed from subtask.
  ❯ This is a subtask.
  ◼ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G❯ Title changed from subtask.
  ✔ This is a subtask.
  ◼ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G❯ Title changed from subtask.
  ✔ This is a subtask.
  ❯ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G❯ Title changed from subtask.
  ✔ This is a subtask.
  ✔ This is a subtask.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G✔ Title changed from subtask.
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "✔ Title changed from subtask.
",
  ],
]
`;